typing/typecore.ml

```diff
@@ -4332,7 +4354,7 @@ and type_application env app_loc (mode : Alloc_mode.t) funct sargs =
           let ty_fun = match td with Tarrow _ -> newty td | _ -> ty_fun in
           let ty_res =
             result_type (!omitted_parameters @ !eliminated_optional_arguments)
```

Comment:
In the presence of modes I don't think this actually produces the right type. I think that we probably need to keep the `eliminated_optional_arguments` in amongst the `omitted_parameters` to maintain their order, so that we can produce the correct type here.

It only affects one very rare error message, so feel free the just mark it as a `FIXME` for now. The error message in question is:
```ocaml
#   let bar (foo : a:int -> ?b:int -> c:int -> 'a) =
    if true then ignore (foo ~b:1 ~a:1 ~c:1 () 1 : unit)
    else foo ~c:1 () ~d:1;;

    Line 4, characters 24-25:
4 |     else foo ~c:1 () ~d:1;;
                            ^
Error: The function applied to this argument has type
         ?b:int -> a:int -> int -> unit
This argument cannot be applied with label ~d
```
Note how the error message reverses the order of the `a` and `b` parameters. I think that is likely to screw up the return modes now. If we keep them in order, then `result_type` will naturally produce the correct type.